source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+    /**
+     * The number of documents that were deleted.
+     */
+    deletedCount: Int64;
+}
+```
+
+#### Summary vs. Verbose Results
+
+Users MUST be able to discern whether a `BulkWriteResult` contains summary or verbose results
+without inspecting the value provided for `verboseResults` in `BulkWriteOptions`. Drivers MUST
+implement this in one of the following ways:
+
+- Expose the `hasVerboseResults` field in `BulkWriteResult` as defined above.
+- Implement the `insertResults`, `upsertResults`, and `deleteResults` fields as optional types and
+  document that they will be unset when `verboseResults` is false.
+- Introduce separate `SummaryBulkWriteResult` and `VerboseBulkWriteResult` types.
+  `VerboseBulkWriteResult` MUST have all of the required fields defined on `BulkWriteResult` above.
+  `SummaryBulkWriteResult` MUST have all of the required fields defined on `BulkWriteResult` above
+  except `insertResults`, `updateResults`, and `deleteResults`.
+
+#### Individual Results
+
+The `InsertOneResult`, `UpdateResult`, and `DeleteResult` classes are the same as or similar to
+types of the same name defined in the [CRUD specification](crud.md). Drivers MUST redefine these
+classes if their existing result classes deviate from the definitions in this specification (e.g.
+if they contain acknowledgement information, which is not applicable for individual bulk write
+operations). Drivers MAY reuse their existing types for these classes if they match the ones
+defined here exactly.
+
+### Exception
+
+```typescript
+class BulkWriteException {
+    /**
+     * A top-level error that occurred when attempting to communicate with the server or execute
+     * the bulk write. This value may not be populated if the exception was thrown due to errors
+     * occurring on individual writes.
+     */
+    error: Optional<Error>;
+
+    /**
+     * Write concern errors that occurred while executing the bulk write. This list may have
+     * multiple items if more than one server command was required to execute the bulk write.
+     */
+    writeConcernErrors: WriteConcernError[];
+
+    /**
+     * Errors that occurred during the execution of individual write operations. This map will
+     * contain at most one entry if the bulk write was ordered.
+     */
+    writeErrors: Map<Int64, WriteError>;
+
+    /**
+     * The results of any successful operations that were performed before the error was
+     * encountered.
+     */
+    partialResult: Optional<BulkWriteResult>;
+}
+```
+
+## Building a `bulkWrite` Command
+
+The `bulkWrite` server command has the following format:
+
+```json
+{
+    "bulkWrite": 1,
+    "ops": <Array>,
+    "nsInfo": <Array>,
+    "errorsOnly": <Boolean>,
+    "ordered": <Boolean>,
+    "bypassDocumentValidation": <Boolean>,
+    "comment": <Bson>,
+    "let": <Document>,
+    ...additional operation-agnostic fields
+}
+```
+
+Drivers SHOULD use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the
```

Comment:
```suggestion
Drivers MUST use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the
```

Suggest preferring "MUST" over "SHOULD" to reduce possible variations in driver implementations. I expect it is easier to relax the spec later if needed.

Preferring "MUST" is also recommended in https://github.com/mongodb/specifications?tab=readme-ov-file#when-in-doubt-use-must-not-should-in-specs